### DISCUSSION TITLE
esp32 pwm temperature fan controller

### DISCUSSION THREAD

**Discussion by montyfert** [arendst/Tasmota#19646] [Q&A] [answered]
hi im trying to make a temperature controlled pwm fan controller,
i have an esp32 with tasmota loaded on it.
gpio14 is pwm going to pwm of fan 
5v feed to esp32 
12v feed to fan 
negatives conected
in console 
pwm 100 sets speed of fan 
i have a  DS18B20 sensor 
will connect  to 3v and ground and gpio ?

i would like to know if it is posible to control the speed of the fan depending on the temperature, 
ie on at 28.5C @ 10%  up in increments to 30c @ 100%  off at 28C
any help would be much apreciated.
thanks 
martin

**Comment by sfromis** [comment]
You could do it with the Berry language, something like this:
```
tasmota.add_rule('DS18B20#Temperature',
  def (temp)
   var speed = temp # make your calculation here
   tasmota.set_timer(0, /->tasmota.cmd('PWM ' + str(speed)))
  end)
```

**Reply by montyfert** [reply]
thanks any chance you could siplify it a bit for me im out of my depth.

**Reply by sfromis** [reply]
It is already simplified.

**Comment by Jason2866** [comment]
Tasmota is not a ready to use solution for all kind of things. It needs tinkering and learning to solve custom scenarios. This is a place to ask specific questions when get stuck. You can't expect anyone will provide a full working solution.

**Reply by montyfert** [reply]
i dont expect anyone to provide a full working solution , just a bit of help.

**Reply by sfromis** [reply]
Yeah, that was what I posted. If you have specific questions, do ask. But first take a look at the [Berry](https://tasmota.github.io/docs/Berry/) language and Tasmota integration.

**Reply by barbudor** [reply, marked as answer]
Tasmota also has a PID feature for managing regulation
It may be possible to use for that use-case : https://tasmota.github.io/docs/PID-Control/

**Reply by montyfert** [reply]
thank you

**Reply by Bond246** [reply]
have you done it with tasmota in the end? 
Im interested in you working solution?

I've done nearly the same with esphome based on that repo https://github.com/patrickcollins12/esphome-fan-controller . But since some weeks my esphome on an ESP32 continuesly reboots which is not a good solution so i'm interested in alternative scenarios.

**Reply by montyfert** [reply]
hi i managed to use tasmota and node red to get it to work.
i used an esp32 relay board 
<img width="1293" height="515" alt="image" src="https://github.com/user-attachments/assets/57c31bb4-863b-4301-a68b-21e8e89330a4" />
to turn on and read the fan speed etc then used node red to match the speed to a temperature and control the fan via mqtt.
<img width="1118" height="291" alt="image" src="https://github.com/user-attachments/assets/4c299a7c-8dba-43e2-af0d-b04c942c4613" />

`[{"id":"b29332228c284f1e","type":"tab","label":"Flow 1","disabled":false,"info":"","env":[]},{"id":"07b880d258892a05","type":"group","z":"b29332228c284f1e","style":{"stroke":"#999999","stroke-opacity":"1","fill":"none","fill-opacity":"1","label":true,"label-position":"nw","color":"#a4a4a4"},"nodes":["9d343cc8c4875c68","d0b84ad87557963c","f8616d7c2df36d18","f51c627033b74ba9","f715a8db609c11ea"],"x":14,"y":59,"w":592,"h":122},{"id":"9d343cc8c4875c68","type":"mqtt out","z":"b29332228c284f1e","g":"07b880d258892a05","name":"pwm1","topic":"cmnd/pwm_fan/pwm1","qos":"","retain":"","respTopic":"","contentType":"","userProps":"","correl":"","expiry":"","broker":"dc0340c9643bd396","x":530,"y":140,"wires":[]},{"id":"d0b84ad87557963c","type":"change","z":"b29332228c284f1e","g":"07b880d258892a05","name":"change","rules":[{"t":"set","p":"payload","pt":"msg","to":"payload.Temperature","tot":"msg"}],"action":"","property":"","from":"","to":"","reg":false,"x":240,"y":140,"wires":[["f8616d7c2df36d18"]]},{"id":"f8616d7c2df36d18","type":"range","z":"b29332228c284f1e","g":"07b880d258892a05","minin":"23.5","maxin":"26","minout":"100","maxout":"1000","action":"clamp","round":true,"property":"payload","name":"","x":390,"y":140,"wires":[["9d343cc8c4875c68"]]},
{"id":"f51c627033b74ba9","type":"comment","z":"b29332228c284f1e","g":"07b880d258892a05","name":"print room temp/fan speed","info":"","x":150,"y":100,"wires":[]},{"id":"f715a8db609c11ea","type":"mqtt in","z":"b29332228c284f1e","g":"07b880d258892a05","name":"esp32 dsb","topic":"tele/SHT3X/SENSOR","qos":"2","datatype":"auto-detect","broker":"dc0340c9643bd396","nl":false,"rap":true,"rh":0,"inputs":0,"x":100,"y":140,"wires":[["d0b84ad87557963c"]]},{"id":"dc0340c9643bd396","type":"mqtt-broker","name":"","broker":"192.168.0.107","port":"1883","clientid":"","autoConnect":true,"usetls":false,"protocolVersion":"4","keepalive":"60","cleansession":true,"autoUnsubscribe":true,"birthTopic":"","birthQos":"0","birthPayload":"","birthMsg":{},"closeTopic":"","closeQos":"0","closePayload":"","closeMsg":{},"willTopic":"","willQos":"0","willPayload":"","willMsg":{},"userProps":"","sessionExpiry":""}]`
